File: src/org/tec/datastructures/linear/CircularList.java
package org.tec.datastructures.linear;

import org.tec.datastructures.nodes.NodoSimple;

public class CircularList<T extends Comparable<T>> {
	private NodoSimple<T> primero;
	private NodoSimple<T> ultimo;
	private int numeroElementos;
	

	public CircularList() {
		this.primero = null;
		this.ultimo = this.primero;
		this.numeroElementos = 0;
		
	}
				
	public boolean isEmpty() {
		return this.primero == null;

	}
	
	
	public NodoSimple<T> getFirst() {
		return primero;
	}

	public void setFirst(NodoSimple<T> primero) {
		this.primero = primero;
	}

	public void insertEnd(T valor) {
		if (this.primero == null) {
			this.primero = new NodoSimple<T>(valor);
			this.primero.setNext(this.primero);
			this.ultimo = this.primero;
		} else {
			NodoSimple<T> actual = new NodoSimple<T>(valor);
			this.ultimo.setNext(actual);
			actual.setNext(this.primero);
			this.ultimo = actual;
		}
		this.numeroElementos++;
			
	}
	
	public void insertFirst(T valor) {
		if(this.primero == null) {
			this.primero = new NodoSimple<T>(valor);
			this.primero.setNext(this.primero);
			this.ultimo = this.primero;
		} else {
			NodoSimple<T> actual = new NodoSimple<T>(valor);
			actual.setNext(this.primero);
			this.primero = actual;
			ultimo.setNext(this.primero);
		}
		this.numeroElementos++;
			
	}
	
		
	public void print() {
		NodoSimple<T> actual = this.primero;
		while (actual != this.ultimo) {
			System.out.println(actual.getValue());
			actual = actual.getNext();
		}
		System.out.println(this.ultimo.getValue());

	}
	

	public NodoSimple<T> search(T buscado) {
		if (this.primero == null)
			return null;
		NodoSimple<T> actual = this.primero.getNext();
		while (actual != this.ultimo.getNext()) {
			if (actual.getValue().compareTo(buscado) == 0) {
				return actual;
			} else {
				actual = actual.getNext();
			}
		}
		if (this.primero.getValue().compareTo(buscado) == 0) {
			return this.ultimo;
		}
		return null;
	}
	
	

	public void remove(T valor) {
		if (search(valor) != null) {
			if (this.primero.getValue() == valor) {
				this.primero = this.primero.getNext();
				this.ultimo.setNext(this.primero);
			} else {
				NodoSimple<T> actual = this.primero;
				while(actual.getNext().getValue() != valor) {
					actual = actual.getNext();
				}
				if (actual.getNext() == this.ultimo) {
                    actual.setNext(this.primero);
                    this.ultimo = actual;
                } else {
                	NodoSimple<T> siguiente = actual.getNext();
                	actual.setNext(siguiente.getNext());	
                }
			}
		}
		this.numeroElementos--;
	}

	public int getSize() {
		return numeroElementos;
	}


}

File: Rider/org.tec.datastructures/Linear/CircularList.cs
using System;
using org.tec.datastructures.Nodes;

namespace org.tec.datastructures.Linear{
    
    public class CircularList<T> where T : IComparable<T>{
        
        private SimpleNode<T> _head;
	
        public CircularList() {
            _head = null;
            
        }
	
        public CircularList(SimpleNode<T> head, SimpleNode<T> next) {
            _head = head;
            if (_head != null) {
                _head.SetNext(next);
            }
		
        }
	
        public SimpleNode<T> GetHead(){
            return _head;
        }
	
	
        public void Append (T value) {
            if (_head == null){
                _head = new SimpleNode<T>(value);
                _head.SetNext(_head);
                return;
            }
            T headCopy = _head.GetValue();
            SimpleNode<T> headNextCopy = _head.GetNext();
		
            _head.SetValue(value);
		
            SimpleNode<T> insert = new SimpleNode<T>(headCopy);
		
            _head.SetNext(insert);
            insert.SetNext(headNextCopy);
		
            _head = insert;
        }
	
	
        public SimpleNode<T> Search(T value) {
            SimpleNode<T> temp = _head;
		
            if (_head == null) {
                return null;
            } 
		
            do {
                if(temp.GetValue().Equals(value)){
                    return temp;
                }
                temp = temp.GetNext();
			
            }while(temp != _head);
		
            return null;
        }

        public void Delete(T value) {
            SimpleNode<T> temp = _head;
		
            if (_head == null) {
                return;
            }
		
            do {
                if(temp.GetValue().Equals(value)) {
                    if (temp == _head) {
                        if (Length() == 1) {
                            _head = null;
                        }
                    }
                    temp.SetValue(temp.GetNext().GetValue());
                    if (temp.GetNext() == _head) {
                        _head = temp;
                    }
                    temp.SetNext(temp.GetNext().GetNext());
                    break;
                }
                temp = temp.GetNext();
			
            }while(temp != _head);
		
        }		
	
	
        public bool IsEmpty(){
            return _head == null;
        }
	
	
        public int Length() {
            SimpleNode<T> temp = _head;
            if (temp == null) {
                return 0;
            }
            int result = 0;
            do {
                result++;
                temp = temp.GetNext();
            }while (temp != _head);
            return result;
        }

        public void Clear() {
            _head = null;
        }
        
        
    }
}